File: c++/recur.cpp
#include <iostream>
#include <tuple>
#include <vector>
#include <bits/stdc++.h>
using namespace std;

//function for getting the sequence length
//recursive version
long collatz(long n)
{
    if(n == 1)
    {
        return 0;
    }
    else
    {
        if(n % 2 == 0)
        {
            return collatz(n/2) + 1;
        }
        else
        {
            return collatz(3*n + 1) + 1;
        }
    }
}

//custom sort for tuples
bool sort2(const tuple<long, long>& a, const tuple<long, long>& b)
{
    return (get<1>(a) < get<1>(b));
}

int main()
{
    //get user input
    cout << "Enter the upper bound: ";
    long n;
    cin >> n;

    //initialize array to hold the largest
    vector<tuple<long, long>> largest;
    largest.push_back(make_tuple(1, 0));

    //main loop
    for(long i = 1; i <= n; i++)
    {
        //get the current sequence length
        long l = collatz(i);
        //get the least significant sequence length of stored values
        tuple<long, long> temp = largest.at(0);

        //if current is more significant
        if(l > get<1>(temp))
        {
            if(largest.size() >= 10)
            {
                //replace if array is holding ten elements
                largest.erase(largest.begin());
            }
            //otherwise add to end
            largest.push_back(make_tuple(i, l));

            //sort by value descending, then by sequence length ascending
            sort(largest.begin(), largest.end());
            sort(largest.begin(), largest.end(), sort2);
            //result is the least significant element at index 0
        }
    }

    //sort by value ascending, then by sequence length descending
    sort(largest.begin(), largest.end());
    reverse(largest.begin(), largest.end());
    sort(largest.begin(), largest.end(), sort2);
    reverse(largest.begin(), largest.end());

    //print the stored values
    for(int i = 0; i < largest.size(); i++)
    {
        cout << get<0>(largest.at(i)) << " " << get<1>(largest.at(i)) << endl;
    }

    return 0;
}

File: julia/recur.jl
#function for getting the sequence length
#recursive version
function collatz(n)
    if n == 1
        return 0
    else
        if n % 2 == 0
            return collatz(n/2) + 1
        else
            return collatz(3*n + 1) + 1
        end
    end
end

#get user input
print("Enter the upper bound: ")
n = parse(Int64, readline())

#initialize array to hold the largest
#used sorted tuples to store values
largest = Tuple{Int64,Int64}[]
push!(largest, (1, 0))

#main loop
for i = 1:n
    #get the current sequence length
    l = collatz(i)
    #get the least significant sequence length of stored values
    temp = largest[1]

    #if current is more significant
    if l > temp[2]
        if length(largest) >= 10
            #replace if array is holding ten elements
            deleteat!(largest, 1)
        end
        #otherwise add to end
        push!(largest, (i, l))

        #sort by value descending, then by sequence length ascending
        sort!(largest)
        sort!(largest, lt = (x, y) -> (x[2] < y[2]))
        #result is the least significant element at index 0
    end
end

#sort by value ascending, then by sequence length descending
sort!(largest)
sort!(largest, lt = (x, y) -> (x[2] > y[2]))

#print the stored values
for i = 1:length(largest)
    println(largest[i][1], " ", largest[i][2])
end
